### PR TITLE
Drop dependency on Q, use native promises

### PR DESCRIPTION
**Pull request by raphinesse**
- Part of https://github.com/apache/cordova/issues/7
- I took care to make a dedicated commit for every change that wasn't a simple search-and-replace
- Apart from the actual removal of Q this fixes a few problems that surfaced thereafter

### CODE REVIEW

File: src/cordova/requirements.js [hunk 1]
```diff
@@ -32,21 +32,18 @@ var knownPlatforms = require('../platforms/platforms');
  *   requirements check results for each platform.
  */
 module.exports = function check_reqs (platforms) {
-    return Q().then(function () {
-        var platforms = cordova_util.preProcessOptions(platforms).platforms; // eslint-disable-line no-use-before-define
+    return Promise.resolve().then(() => {
```
Comment (dpogue) [inline src/cordova/requirements.js]: Is there a reason to wrap this in a promise, given that the inner block is already returning `Promise.all`?

File: src/cordova/restore-util.js [hunk 2]
```diff
@@ -111,7 +110,7 @@ function installPlatformsFromConfigXML (platforms, opts) {
 
         // No platforms to restore from either config.xml or package.json.
         if (comboArray.length <= 0) {
-            return Q('No platforms found in config.xml or package.json. Nothing to restore');
+            return Promise.resolve('No platforms found in config.xml or package.json. Nothing to restore');
```
Comment (dpogue) [inline src/cordova/restore-util.js]: This pattern seems awkward to me, since it feels more like these should be emitted as logging events.
Does this function always resolve with a log message string, or with an array of the added platforms? Currently it does both and it would be better in terms of API to be consistent with the resolved value type.

File: src/hooks/HooksRunner.js [hunk 3]
```diff
@@ -45,7 +44,7 @@ function HooksRunner (projectRoot) {
  */
 HooksRunner.prototype.fire = function fire (hook, opts) {
     if (isHookDisabled(opts, hook)) {
-        return Q('hook ' + hook + ' is disabled.');
+        return Promise.resolve('hook ' + hook + ' is disabled.');
```
Comment (dpogue) [inline src/hooks/HooksRunner.js]: This method resolves with a string, or throws an error (not rejects), or resolves with the result of running scripts, or rejects with error from running scripts? :grimacing:

File: src/cordova/requirements.js (hunk 1 above)
Comment (raphinesse) [reply src/cordova/requirements.js]: It's a poor-man's [`Promise.try`](http://cryto.net/~joepie91/blog/2016/05/11/what-is-promise-try-and-why-does-it-matter/).

In a nutshell: It ensures that this function  __always__ returns a `Promise`. Even if anything we do would cause an error to be thrown, thanks to the wrapping it will be converted to a rejection instead of bubbling up the call stack.

In this case, suppose `preProcessOptions(platforms)` would return `undefined`. Then our accessing the result's `platforms` property would result in an error to be thrown.

Comment (dpogue) [reply src/cordova/requirements.js]: Ahh, I actually just left a comment on the cordova-fetch PR referencing `Promise.try`. Since we seem to use this pattern frequently, maybe it makes sense to depend on something like https://github.com/tc39/proposal-promise-try (although that isn't published to npm :disappointed:)?

Comment (raphinesse) [reply src/cordova/requirements.js]: Yeah, sorry for doing a ELI5 here. And I agree with you regarding the lib: I just suggested the exact same thing in the cordova-fetch PR :grin:

**Comment by raphinesse** [comment]
> we can make note of those other cleanup tasks somewhere

Could you take care of that please @dpogue? :pray:

**Comment by piotr-cz** [comment]
Along with this change, docs for [Hooks: Sample Usage](https://cordova.apache.org/docs/en/latest/guide/appdev/hooks/#sample-usage) should be updated

**Comment by raphinesse** [comment]
@piotr-cz Agreed. Actually the whole page is quite dated. I will try to fix that in the next few days.

**Comment by raphinesse** [comment]
@piotr-cz See https://github.com/apache/cordova-docs/pull/958
